main.tf

```diff
@@ -0,0 +1,166 @@
+
+  backend "s3" {
+    bucket               = "ccao-terraform-state-us-east-1"
+    key                  = "workspaces/default/terraform.tfstate"
+    region               = "us-east-1"
+    workspace_key_prefix = "workspaces"
+  }
+}
+
+provider "aws" {
+  region = "us-east-1"
+}
+
+locals {
+  s3_prefix                = "scripts/sales-val"
+  s3_bucket_data_warehouse = terraform.workspace == "prod" ? "ccao-data-warehouse-us-east-1" : aws_s3_bucket.data_warehouse[0].id
+  s3_bucket_glue_assets    = terraform.workspace == "prod" ? "ccao-glue-assets-us-east-1" : aws_s3_bucket.glue_assets[0].id
+  glue_job_name            = "sales_val_flagging${terraform.workspace == "prod" ? "" : "_${terraform.workspace}"}"
+  glue_crawler_name        = "ccao-data-warehouse-sale-crawler${terraform.workspace == "prod" ? "" : "-${terraform.workspace}"}"
+}
+
+variable "iam_role_arn" {
+  type        = string
+  description = "IAM role ARN to use to run the job"
+  sensitive   = true
+  nullable    = false
+}
+
+variable "commit_sha" {
+  type        = string
+  description = <<EOF
+    "Hash of the current git commit, to use for versioning the model. "
+    "Generate with `git rev-parse --short HEAD`"
+  EOF
+  nullable    = false
+}
+
+resource "aws_s3_bucket" "glue_assets" {
+  # Prod buckets are managed outside this config
+  count         = terraform.workspace == "prod" ? 0 : 1
+  bucket        = "ccao-glue-assets-${terraform.workspace}-us-east-1"
+  force_destroy = true
+}
+
+resource "aws_s3_bucket_public_access_block" "glue_assets" {
+  count  = terraform.workspace == "prod" ? 0 : 1
+  bucket = local.s3_bucket_glue_assets
+
+  block_public_acls       = true
+  block_public_policy     = true
+  ignore_public_acls      = true
+  restrict_public_buckets = true
+}
+
+resource "aws_s3_bucket_versioning" "glue_assets" {
+  count  = terraform.workspace == "prod" ? 0 : 1
+  bucket = local.s3_bucket_glue_assets
+
+  versioning_configuration {
+    status = "Enabled"
+  }
+}
+
+resource "aws_s3_bucket" "data_warehouse" {
+  count         = terraform.workspace == "prod" ? 0 : 1
+  bucket        = "ccao-data-warehouse-${terraform.workspace}-us-east-1"
+  force_destroy = true
+}
+
+resource "aws_s3_bucket_public_access_block" "data_warehouse" {
+  count  = terraform.workspace == "prod" ? 0 : 1
+  bucket = local.s3_bucket_data_warehouse
+
+  block_public_acls       = true
+  block_public_policy     = true
+  ignore_public_acls      = true
+  restrict_public_buckets = true
+}
+
+resource "aws_s3_bucket_versioning" "data_warehouse" {
+  count  = terraform.workspace == "prod" ? 0 : 1
+  bucket = local.s3_bucket_data_warehouse
+
+  versioning_configuration {
+    status = "Enabled"
+  }
+}
+
+resource "aws_s3_object" "sales_val_flagging" {
+  bucket = local.s3_bucket_glue_assets
+  key    = "${local.s3_prefix}/sales_val_flagging.py"
+  source = "${path.module}/glue/sales_val_flagging.py"
+  etag   = filemd5("${path.module}/glue/sales_val_flagging.py")
```

Comment:
Combined, these two attributes allow us to upload a new version of the script to S3 whenever its contents change. [`source`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_object#source) indicates the path to the file in the local filesystem that should be uploaded to S3, while [`etag`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_object#etag) defines a value (in this case, the md5 hash of the file) that determines whether the resource should be updated or not -- i.e. if the hash changes, the file must be uploaded.